File: android-app/src/main/java/fr/cph/chicago/entity/dto/BusArrivalRouteDTO.java
package fr.cph.chicago.entity.dto;

import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import fr.cph.chicago.entity.BusArrival;

/**
 * @author cpharmant
 */
public class BusArrivalRouteDTO extends TreeMap<String, Map<String, List<BusArrival>>> {
    // route => { bound => BusArrival }

    public void addBusArrival(@NonNull final BusArrival busArrival) {
        if (containsKey(busArrival.getRouteId())) {
            final Map<String, List<BusArrival>> tempMap = get(busArrival.getRouteId());
            if (tempMap.containsKey(busArrival.getRouteDirection())) {
                final List<BusArrival> arrivals = tempMap.get(busArrival.getRouteDirection());
                arrivals.add(busArrival);
            } else {
                final List<BusArrival> arrivals = new ArrayList<>();
                arrivals.add(busArrival);
                tempMap.put(busArrival.getRouteDirection(), arrivals);
            }
        } else {
            final Map<String, List<BusArrival>> tempMap = new TreeMap<>();
            final List<BusArrival> arrivals = new ArrayList<>();
            arrivals.add(busArrival);
            tempMap.put(busArrival.getRouteDirection(), arrivals);
            put(busArrival.getRouteId(), tempMap);
        }
    }
}

File: android-app/src/main/java/fr/cph/chicago/entity/dto/BusArrivalRouteDTO.kt
package fr.cph.chicago.entity.dto

import java.util.ArrayList
import java.util.TreeMap

import fr.cph.chicago.entity.BusArrival

/**
 * @author cpharmant
 */
class BusArrivalRouteDTO : TreeMap<String, Map<String, List<BusArrival>>>() {
    // route => { bound => BusArrival }

    fun addBusArrival(busArrival: BusArrival) {
        // FIXME kotlin
        /*if (containsKey(busArrival.routeId)) {
            val tempMap = get(busArrival.routeId)
            if (tempMap!!.containsKey(busArrival.routeDirection)) {
                val arrivals: List<BusArrival>? = tempMap[busArrival.routeDirection.orEmpty()]
                arrivals.add(busArrival)
            } else {
                val arrivals = ArrayList<BusArrival>()
                arrivals.add(busArrival)
                tempMap.put(busArrival.getRouteDirection(), arrivals)
            }
        } else {
            val tempMap = TreeMap<String, List<BusArrival>>()
            val arrivals = ArrayList<BusArrival>()
            arrivals.add(busArrival)
            tempMap.put(busArrival.getRouteDirection(), arrivals)
            put(busArrival.routeId.orEmpty(), tempMap)
        }*/
    }
}
